plugins/modules/mso_schema_site_service_graph_node.py

```diff
@@ -0,0 +1,293 @@
+
+    mso.existing = {}
+    service_graph_idx = None
+
+    # Get Service Graph
+    service_graph_ref = mso.service_graph_ref(schema_id=schema_id, template=template, service_graph=service_graph)
+    service_graph_refs = [f.get('serviceGraphRef') for f in schema_obj.get('sites')[site_idx]['serviceGraphs']]
+    if service_graph is not None and service_graph_ref in service_graph_refs:
+        service_graph_idx = service_graph_refs.index(service_graph_ref)
+        mso.existing = schema_obj.get('sites')[site_idx]['serviceGraphs'][service_graph_idx]
+
+    if state == 'query':
+        if service_graph is None:
+            mso.existing = schema_obj.get('sites')[site_idx]['serviceGraphs']
+        if service_graph is not None and service_graph_idx is None:
+            mso.fail_json(msg="Service Graph '{service_graph}' not found".format(service_graph=service_graph))
+        mso.exit_json()
+
+    service_graphs_path = '/sites/{0}/serviceGraphs/-'.format(site_template)
+    service_graph_path = '/sites/{0}/serviceGraphs/{1}'.format(site_template, service_graph)
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        if mso.existing:
+            mso.sent = mso.existing = {}
+            ops.append(dict(op='remove', path=service_graph_path))
+
+    elif state == 'present':
+        devices_payload = []
+        node_device_type = []
+        device_types = {}
+        device_number = 0
+        service_graphs = list(templates[template_idx]['serviceGraphs'])
+        service_nodes_list = [r['serviceNodes'] for r in service_graphs if r.get('name') == service_graph]
+        service_nodes_list_types = [k.get('name') for k in service_nodes_list[0]]
+        user_number_devices = len(devices)
+        number_service_nodes = len(service_nodes_list[0])
+        if user_number_devices != number_service_nodes:
+            mso.fail_json(msg="Service Graph '{0}' has '{1}' service node type(s) but '{2}' service node type(s) were given for the device"
+                          .format(service_graph, number_service_nodes, user_number_devices))
+        # Populate dict for service node type and device type
+        for node_type in service_nodes_list_types:
+            if node_type == 'firewall':
+                node_device_type.append('FW')
+            elif node_type == 'load-balancer':
+                node_device_type.append('ADC')
+            else:
+                node_device_type.append('OTHERS')
+
+        # Populate dict for device types and an index number
+        keys_number_devices = range(user_number_devices)
+        for number in keys_number_devices:
+            device_types[number] = node_device_type[number]
+
+        if devices is not None:
+            for device in devices:
+                device_name = device.get('name')
+                query_device_data = mso.query_service_node_device_types(site_id, tenant, device_types[device_number], device_name)
```

Comment:
create a lookup_service_node_device function which return the device if found or error out.

Reply:
ok